audio_service/android/src/main/java/com/ryanheise/audioservice/AudioServicePlugin.java

```diff
@@ -49,197 +49,212 @@
-    }
-    public static synchronized FlutterEngine getFlutterEngine(Context context) {
-        FlutterEngine flutterEngine = FlutterEngineCache.getInstance().get(flutterEngineId);
-        if (flutterEngine == null) {
-            // XXX: The constructor triggers onAttachedToEngine so this variable doesn't help us.
-            // Maybe need a boolean flag to tell us we're currently loading the main flutter engine.
-            flutterEngine = new FlutterEngine(context.getApplicationContext());
-            flutterEngine.getDartExecutor().executeDartEntrypoint(DartExecutor.DartEntrypoint.createDefault());
-            FlutterEngineCache.getInstance().put(flutterEngineId, flutterEngine);
-        }
-        return flutterEngine;
-    }
-
-    public static void disposeFlutterEngine() {
-        FlutterEngine flutterEngine = FlutterEngineCache.getInstance().get(flutterEngineId);
-        if (flutterEngine != null) {
-            flutterEngine.destroy();
-            FlutterEngineCache.getInstance().remove(flutterEngineId);
-        }
-    }
-
+    private static final String DEFAULT_ENGINE_ID = "audio_service_engine";
     private static final String CHANNEL_CLIENT = "com.ryanheise.audio_service.client.methods";
     private static final String CHANNEL_HANDLER = "com.ryanheise.audio_service.handler.methods";
 
-    private static Context applicationContext;
-    private static final Set<ClientInterface> clientInterfaces = new HashSet<>();
-    private static ClientInterface mainClientInterface;
-    private static AudioHandlerInterface audioHandlerInterface;
-    private static final long bootTime;
-    private static Result configureResult;
-
-    static {
-        bootTime = System.currentTimeMillis() - SystemClock.elapsedRealtime();
+    static FlutterEngineFactory.EngineConfig engineConfig;
+    @Nullable private static FlutterEngineFactory engineFactory;
+    /**
+     * Sets a custom factory for {@link FlutterEngine}.
+     *
+     * <p>By default the engine will be created with default ID and
+     * will be pre-warmed and managed internally.
+     */
+    public static void setEngineFactory(@Nullable FlutterEngineFactory value) {
+        engineFactory = value;
     }
 
-    static AudioHandlerInterface audioHandlerInterface() throws Exception {
-        if (audioHandlerInterface == null) throw new Exception("Background audio task not running");
-        return audioHandlerInterface;
-    }
+    /** Factory callback for {@link FlutterEngine}. */
+    public interface FlutterEngineFactory {
+        /** Runs the factory */
+        EngineConfig create();
+
+        /** Config of the {@link FlutterEngine} for the plugin. */
+        class EngineConfig {
+            private String id;
+            private FlutterEngine engine;
+
+            /**
+             * Creates config with ID.
+             * Engine will be pre-warmed and managed internally.
+             */
+            public EngineConfig(String id) {
```

Comment:
is there a use-case for setting a custom ID? i kept this from the previous plugin's code, but is there a use-case for that? a conflict in the default engine ID is **extremely** unlikely

Reply:
In a typical app you'd be right, although it's possible that an existing Android app may incrementally convert some of its activities over to flutter and use the embedding API to do so. For that, they will probably manage their own flutter engines, and may have some infrastructure code in place dealing with the flutter engine cache (which is where the id gets used), so I just wanted to make it easy to do that (and it was extremely cheap to add a setter for the ID).